chia/full_node/subscriptions.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional, Set
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+
+def _get_or_add_set(c: Dict[bytes32, Set[bytes32]], key: bytes32) -> Set[bytes32]:
+    ret: Optional[Set[bytes32]] = c.get(key)
+    if ret is not None:
+        return ret
+
+    ret2: Set[bytes32] = set()
+    c[key] = ret2
+    return ret2
+
+
+# The PeerSubscriptions class is essentially a multi-index container. It can be
+# indexed by peer_id, coin_id and puzzle_hash.
+@dataclass
+class PeerSubscriptions:
+    # TODO: use NewType all over to describe these various uses of the same types
+    # Puzzle Hash : Set[Peer ID]
+    _coin_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Puzzle Hash : Set[Peer ID]
+    _ph_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: Set[Coin ids]
+    _peer_coin_ids: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: Set[puzzle_hash]
+    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
+    # Peer ID: subscription count
+    _peer_sub_counter: Dict[bytes32, int] = field(default_factory=dict, init=False)
+
+    def has_ph_sub(self, ph: bytes32) -> bool:
+        return ph in self._ph_subscriptions
+
+    def has_coin_sub(self, coin_id: bytes32) -> bool:
+        return coin_id in self._coin_subscriptions
+
+    def add_ph_subscriptions(self, peer_id: bytes32, phs: List[bytes32], max_items: int) -> None:
+        puzzle_hash_peers = _get_or_add_set(self._peer_puzzle_hash, peer_id)
+
+        existing_sub_count = self._peer_sub_counter.get(peer_id)
+        if existing_sub_count is None:
+            self._peer_sub_counter[peer_id] = 0
+            existing_sub_count = 0
+
+        # if we've reached the limit on number of subscriptions, just bail
+        if existing_sub_count >= max_items:
+            return
+
+        # decrement this counter as we go, to know if we've hit the limit of
+        # number of subscriptions
+        subscriptions_left = max_items - existing_sub_count
+
+        for ph in phs:
+            ph_sub: Optional[Set[bytes32]] = self._ph_subscriptions.get(ph)
+            if ph_sub is None:
+                ph_sub = set()
+                self._ph_subscriptions[ph] = ph_sub
+            elif peer_id in ph_sub:
```

Comment:
we should add some logging for the return paths, will help people with lots of coins/ph to understand why/if they don't get all their coins when syncing from a local node

Reply:
good point.